### PR TITLE
Document persistence options

### PR DESCRIPTION
**Pull request by appgurueu**
Elaborates on serialization options and provides a persistence option overview

### CODE REVIEW

File: doc/persistence.adoc [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+WARNING: The JSON serializer is furthermore limited by its recursion depth: Only a recursion depth up to *16* is allowed. Very nested data structures can't be (de)serialized. Circular references will cause the JSON serializer to recurse until the maximum recursion depth is exceeded.
+
+TIP: Use the `json = assert(minetest.write_json(data))` idiom to not silently ignore errors when serializing.
+
+==== `minetest.parse_json(json, [nullvalue])`
+
+If `json` is valid JSON:: Returns the Lua value represented by the JSON string `json`, with `null` values deserialized to `nullvalue` (which defaults to `nil`).
+Else:: Returns `nil` and logs an error.
+
+TIP: If you must use JSON (to interact with a web interface, for instance), consider using your JSON library of choice; Lua-only implementations are available.
```
Comment (benrob0329) [inline doc/persistence.adoc:54]: Can you give an example of one which solves the issues presented by the Minetest implementation?

File: doc/persistence.adoc [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+=== Lua
+
+Minetest alternatively offers a serializer implemented in Lua, which serializes to Lua source code, available as  and `minetest.deserialize`. The following values are supported:
```
Comment (benrob0329) [inline doc/persistence.adoc:58]:
```suggestion
Minetest alternatively offers a serializer implemented in Lua, which serializes to Lua source code, available as `minetest.serialize` and `minetest.deserialize`. The following values are supported:
```

File: doc/persistence.adoc [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+* `nil`
+* booleans
+* numbers
+* strings
+* tables, including circular references
+* functions: *not recommended*, require mod security to be disabled
+** Lua-implementation specific bytecode: Functions serialized by PUC Lua won't work on LuaJIT and vice versa
+** Uses `string.dump` internally: Functions may not have upvalues
```
Comment (benrob0329) [inline doc/persistence.adoc:67]: This isn't super clear on what kind of function serialization is supported. Both bytecode and non-bytecode versions of functions can be deserialized (assuming the safe flag is false), but only non-bytecode functions can be serialized (iirc).

File: doc/persistence.adoc [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+Userdata objects (like `ItemStack` for example) aren't supported. Threads aren't supported either.
```
Comment (benrob0329) [inline doc/persistence.adoc:69]:
```suggestion
Userdata objects (like `ItemStack` for example) aren't supported. Threads (coroutines) aren't supported either.
```

File: doc/persistence.adoc [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+* `minetest.cpdir`
+* `minetest.mvdir`
+* `minetest.get_dir_list`
+
+If you want to mitigate the risk of data loss, you can use `minetest.safe_file_write` when rewriting files.
+// TODO document FS helpers thoroughly elsewhere
+
+==== Configuration files
+
+The `Settings` object allows you to operate on configuration files, getting & setting key-value entries and saving the file. The main `Settings` object `minetest.settings` can be used to persist a few settings "globally" - bleeding everywhere. This is horribly abused by the mainmenu to store stuff like the last selected game. Don't be like the mainmenu. Configuration files are presumably easy to edit for users, but so is Lua.
```
Comment (benrob0329) [inline doc/persistence.adoc:150]: This feels a little _too_ opinionated to me. I think that the configuration interface is fine if used properly, and is likely a lot faster for the simple k/v storage it's supposed to be than serialization. Noting that the `minetest.settings` object is global is good, even separating that part out into something like `CAUTION: This sets global settings, don't set global settings.` might be good.

**Review by benrob0329 — CHANGES_REQUESTED** [review]

File: doc/persistence.adoc (hunk 1 above)
Comment (appgurueu) [reply doc/persistence.adoc:54]: I have implemented my own and didn't want to mention that. https://github.com/grafi-tt/lunajson/ could be mentioned here.

File: doc/persistence.adoc [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+* functions: *not recommended*, require mod security to be disabled
+** Lua-implementation specific bytecode: Functions serialized by PUC Lua won't work on LuaJIT and vice versa
+** Uses `string.dump` internally: Functions may not have upvalues
+
+Userdata objects (like `ItemStack` for example) aren't supported. Threads (coroutines) aren't supported either.
+
+==== `minetest.serialize(data)`
+
+Serializes `data`, which may be a value of any of the above types. Returns a Lua chunk in `string` form that returns `data` if ran. Will error with `Can't serialize data of type <type>` if it encounters an unsupported type.
+
```
Comment (sfan5) [inline doc/persistence.adoc:74]: should mention "Note that functions will be dumped to bytecode by this function but will always error upon attempting to deserialize them (unless mod security is disabled), so this is not useful for that purpose."

**Comment by appgurueu** [comment]
Needs to be updated as of https://github.com/minetest/minetest/pull/11427.

**Review by wsor4035 — APPROVED** [review]

**Pull request merged**